### PR TITLE
modcc: AVX512 vectorisation backend

### PR DESCRIPTION
**Pull request by vkarak**
Basic features:

* Compile with `-t avx512`
* Automatically set up by CMake if `USE_OPTIMIZED_KERNELS` is on and `VECTORIZE_TARGET` is set to `KNL`
* Generic SIMD printer that contacts a SIMD backend for emitting the actual SIMD intrinsics

Note: tested with intel compiler only.

### CODE REVIEW

**Comment by bcumming** [comment]
I get compilation errors, specifically within the CUDA printer:
```
[  1%] Building CXX object modcc/CMakeFiles/compiler.dir/cudaprinter.cpp.o
/home/bcumming/nestmc-proto/modcc/cudaprinter.cpp: In constructor ‘CUDAPrinter::CUDAPrinter(Module&, bool)’:
/home/bcumming/nestmc-proto/modcc/cudaprinter.cpp:461:46: error: use of deleted function ‘CPrinter::CPrinter(const CPrinter&)’
             auto cprinter = CPrinter(*module_);
                                              ^
In file included from /home/bcumming/nestmc-proto/modcc/cudaprinter.cpp:3:0:
/home/bcumming/nestmc-proto/modcc/cprinter.hpp:9:7: note: ‘CPrinter::CPrinter(const CPrinter&)’ is implicitly deleted because the default definition would be ill-formed:
 class CPrinter : public Visitor {
       ^~~~~~~~
/home/bcumming/nestmc-proto/modcc/cprinter.hpp:9:7: error: use of deleted function ‘TextBuffer::TextBuffer(const TextBuffer&)’
In file included from /home/bcumming/nestmc-proto/modcc/cprinter.hpp:6:0,
                 from /home/bcumming/nestmc-proto/modcc/cudaprinter.cpp:3:
/home/bcumming/nestmc-proto/modcc/textbuffer.hpp:7:7: note: ‘TextBuffer::TextBuffer(const TextBuffer&)’ is implicitly deleted because the default definition would be ill-formed:
 class TextBuffer {
       ^~~~~~~~~~
/home/bcumming/nestmc-proto/modcc/textbuffer.hpp:7:7: error: use of deleted function ‘std::__cxx11::basic_stringstream<_CharT, _Traits, _Alloc>::basic_stringstream(const std::__cxx11::basic_stringstream<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]’
In file included from /home/bcumming/nestmc-proto/modcc/cprinter.hpp:3:0,
                 from /home/bcumming/nestmc-proto/modcc/cudaprinter.cpp:3:
/usr/include/c++/6.2.1/sstream:721:7: note: declared here
       basic_stringstream(const basic_stringstream&) = delete;
```

**Comment by vkarak** [comment]
I've just tried it with gcc 5.3.0 and I didn't get any error. Let me try with a more recent one.

**Comment by vkarak** [comment]
I've just compiled it with gcc 6.2 and I'm not getting any error apart from some warnings. This is my configuration:

```
 CMAKE_BUILD_TYPE                 RelWithDebInfo                                                                   
 CMAKE_INSTALL_PREFIX             /usr/local                                                                       
 MODCC_BIN                        MODCC_BIN-NOTFOUND                                                               
 NMC_AUTO_RUN_MODCC_ON_CHANGES    ON                                                                               
 NMC_BUILD_VALIDATION_DATA        OFF                                                                              
 NMC_DISTRIBUTED_MODEL            serial                                                                           
 NMC_SYSTEM_TYPE                  Generic                                                                          
 NMC_THREADING_MODEL              tbb                                                                              
 NMC_USE_OPTIMIZED_KERNELS        OFF                                                                              
 NMC_VALIDATION_DATA_DIR          /users/karakasv/Devel/nestmc-proto/validation/data                               
 NMC_VECTORIZE_TARGET             none                                                                             
 NMC_WITH_ASSERTIONS              ON                                                                               
 NMC_WITH_CUDA                    OFF                                                                              
 NMC_WITH_PROFILING               ON                                                                               
 NMC_WITH_TRACE                   OFF                                                                              
 unwind_library_generic           unwind_library_generic-NOTFOUND                                                  
 unwind_library_target            unwind_library_target-NOTFOUND                                                   
```

**Comment by bcumming** [comment]
Nice work!

File: mechanisms/CMakeLists.txt [hunk 1]
```diff
@@ -12,11 +12,16 @@ set(mod_srcdir "${CMAKE_CURRENT_SOURCE_DIR}/mod")
 
 set(mech_dir "${CMAKE_CURRENT_SOURCE_DIR}/multicore")
 file(MAKE_DIRECTORY "${mech_dir}")
+if(NMC_VECTORIZE_TARGET STREQUAL "KNL" AND NMC_USE_OPTIMIZED_KERNELS)
```
Comment (vkarak) [inline mechanisms/CMakeLists.txt]: > Both the true and false branches perform set(modcc_target "cpu"). Do you mean to set(modcc_target avx512) if this is true?

This branch should be `set(modcc_target avx512)`. That's a mistake that slipped in when I was comparing the intrinsics vs. the original optimised versions.

> How about we remove the NMC_USE_OPTIMIZED_KERNELS option from CMakeLists, and assume that the user is requesting an optimized build if they request a vectorized back end.

That's a valid assumption I think.

**Pull request merged**